### PR TITLE
Equal Opportunity implementation + sample notebooks

### PR DESCRIPTION
**Pull request by hannanabdul55**
## What's there in this PR? 
- Initial implementation of Equal Opportunity, which is a slightly modified version of Equalized Odds (specifically for `y=1` case only). 
- Sample notebooks for Demographic Parity constraint evaluation on Law School dataset. Check these notebooks [[1]](https://github.com/UMass-Responsible-AI/fairlearn/blob/master/notebooks/Exp%20Gradient%20for%20DP(adultl).ipynb) [[2]](https://github.com/UMass-Responsible-AI/fairlearn/blob/master/notebooks/Exp%20Gradient%20for%20DP(law%20school).ipynb) [[3]](https://github.com/UMass-Responsible-AI/fairlearn/blob/master/notebooks/Exp%20Gradient%20for%20DP.ipynb).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -114,3 +114,10 @@ docbuild/
 
 # generated perf test scripts
 test/perf/scripts/**
+
+# idea specific stuff
+.idea/**/workspace.xml
```
Comment (riedgar-ms) [inline .gitignore]: Can we just ignore the entire `.idea` directory?

File: fairlearn/reductions/_moments/conditional_selection_rate.py [hunk 2]
```diff
@@ -145,6 +157,49 @@ def load_data(self, X, y, **kwargs):
...
+                          **kwargs)
+
+class EqualOpportunity(ConditionalSelectionRate):
+    short_name = "EqualOpportunity"
+
+    def load_data(self, X, y, **kwargs):
+        """Load the specified data into the object."""
+        super().load_data(X, y,
+                          event=pd.Series(y).apply(lambda y: _LABEL + "=" + str(y)),
+                          filter_value=0,
```
Comment (riedgar-ms) [inline fairlearn/reductions/_moments/conditional_selection_rate.py]: One small tweak: can we make the 'good value' a parameter to the constructor or this method? That might fold into @MiroDudik 's API review, though

File: fairlearn/reductions/_moments/conditional_selection_rate.py [hunk 3]
```diff
@@ -33,13 +34,21 @@ def default_objective(self):
...
+    def load_data(self, X, y, event=None, filter_value=None, **kwargs):
         """Load the specified data into this object.
 
         This adds a column `event` to the `tags` field.
         """
         super().load_data(X, y, **kwargs)
         self.tags[_EVENT] = event
+        if filter_value is not None:
+            self.dropped_tags = self.tags.where(self.tags['label'] == 0).dropna()
+            self.tags = self.tags.where(self.tags['label'] == 1).dropna()
```
Comment (riedgar-ms) [inline fairlearn/reductions/_moments/conditional_selection_rate.py]: Should these two equalities be checking against `filter_value` ?

File: fairlearn/reductions/_moments/conditional_selection_rate.py [hunk 4]
```diff
@@ -5,6 +5,7 @@
 from .moment import ClassificationMoment
 from .moment import _GROUP_ID, _LABEL, _PREDICTION, _ALL, _EVENT, _SIGN
 from .error_rate import ErrorRate
+import numpy as np
```
Comment (riedgar-ms) [inline fairlearn/reductions/_moments/conditional_selection_rate.py]: Is this for `dropna()` ?

File: .gitignore (hunk 1 above)
Comment (hannanabdul55) [reply .gitignore]: Yeah, sure I will ignore the idea folder completely.

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 2 above)
Comment (hannanabdul55) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: Yes, this was the original intention of the flag, but I also need the 'bad value' then so that they can be filtered out. But I think there is a workaround for that which I can do with some pandas API.  I will try it out. Thank you!

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 3 above)
Comment (hannanabdul55) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: Yes, that is correct. I will make that change too

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 4 above)
Comment (hannanabdul55) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: No, this can be removed. This was originally added because of some NumPy array manipulation that I was trying for the DataFrame, which I am not using now.

File: fairlearn/reductions/_moments/conditional_selection_rate.py [hunk 5]
```diff
@@ -145,6 +157,49 @@ def load_data(self, X, y, **kwargs):
         """Load the specified data into the object."""
         super().load_data(X, y, event=_ALL, **kwargs)
 
+class EqualOpportunity(ConditionalSelectionRate):
```
Comment (romanlutz) [inline fairlearn/reductions/_moments/conditional_selection_rate.py]: As mentioned earlier in person, we may want to make this True Positive Parity instead.

File: fairlearn/reductions/_moments/conditional_selection_rate.py [hunk 6]
```diff
@@ -145,6 +157,49 @@ def load_data(self, X, y, **kwargs):
...
+    r"""Implementation of Equalized Odds as a moment.
```
Comment (romanlutz) [inline fairlearn/reductions/_moments/conditional_selection_rate.py]: Documentation needs update

File: fairlearn/reductions/_moments/conditional_selection_rate.py [hunk 7]
```diff
@@ -145,6 +157,49 @@ def load_data(self, X, y, **kwargs):
...
+    the number of unique values for the sensitive feature, multiplied by
+    the number of unique values of the `Y` array, multiplied by two (for
+    the Lagrange multipliers for positive and negative constraints).
+
+    With these definitions, the :meth:`signed_weights` method
+    will calculate the costs according to Example 4 of
+    `Agarwal et al. (2018) <https://arxiv.org/abs/1803.02453>`_.
+    """
+
+    short_name = "EqualizedOdds"
```
Comment (romanlutz) [inline fairlearn/reductions/_moments/conditional_selection_rate.py]: please update the short_name as well

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 5 above)
Comment (romanlutz) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: Wait, did you name both of them EqualOpportunity, i.e. did you rename EqualizedOdds as well?

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 6 above)
Comment (romanlutz) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: unless this is meant to be Equalized Odds. The class name threw me off

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 7 above)
Comment (romanlutz) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: unless this is indeed EO. The class name threw me off

File: fairlearn/reductions/_moments/conditional_selection_rate.py [hunk 8]
```diff
@@ -145,6 +157,49 @@ def load_data(self, X, y, **kwargs):
...
+
+    def load_data(self, X, y, **kwargs):
+        """Load the specified data into the object."""
+        super().load_data(X, y,
+                          event=pd.Series(y).apply(lambda y: _LABEL + "=" + str(y)),
+                          **kwargs)
+
+class EqualOpportunity(ConditionalSelectionRate):
```
Comment (romanlutz) [inline fairlearn/reductions/_moments/conditional_selection_rate.py]: Can we add a short doc section like with EqualizedOdds? Should be fairly similar, and in fact you can refer to Equalized ODds as well

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 2 above)
Comment (romanlutz) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: I agree with Richard here. If you know the "good value" then the "bad value" is simply the other one (using some sort of unique/set magic to determine that). This may come in handy if/when we decide to move to multi-class classification, because suddenly it's not one or the other, but one of n, and then specifying the "good value" is way easier than specifying n-1 "bad values".

File: notebooks/Binary Classification on COMPAS dataset.ipynb [hunk 9]
```diff
@@ -19,9 +19,30 @@
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 1,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
```
Comment (romanlutz) [inline notebooks/Binary Classification on COMPAS dataset.ipynb]: Before committing notebooks we always clear the output. That should reduce the number of changes to a minimum here :-)

File: test/unit/reductions/exponentiated_gradient/test_exponentiatedgradient_smoke.py [hunk 10]
```diff
@@ -103,7 +109,7 @@ def test_smoke(self, testdata):
 
     def test_simple_fit_predict(self):
         estimator = LeastSquaresBinaryClassifierLearner()
-        constraints = DemographicParity()
+        constraints = EqualOpportunity()
```
Comment (romanlutz) [inline test/unit/reductions/exponentiated_gradient/test_exponentiatedgradient_smoke.py]: You may add another test case, but please leave the existing ones in place :-)

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 7 above)
Comment (hannanabdul55) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: Oops yeah, good catch sorry about this. I think I messed up a merge before raising this PR because of which it got added twice!

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 2 above)
Comment (hannanabdul55) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: I have made this change. Looks cleaner than the current implementation for sure, but I don't think that would work for the multi-class classification.

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 5 above)
Comment (hannanabdul55) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: Fixing all this in the next commit. 👍

File: test/unit/reductions/exponentiated_gradient/test_exponentiatedgradient_smoke.py (hunk 10 above)
Comment (hannanabdul55) [reply test/unit/reductions/exponentiated_gradient/test_exponentiatedgradient_smoke.py]: Sorry about this. Yes, I will do this.

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 6 above)
Comment (hannanabdul55) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: Sorry about that. I think all this should be resolved now!

File: fairlearn/reductions/_moments/conditional_selection_rate.py [hunk 11]
```diff
@@ -146,6 +159,43 @@ def load_data(self, X, y, **kwargs):
...
+    fraction of the samples corresponding to each unique value of `y = 1` in
+    the `Y` array.
+
+    The `index` MultiIndex will have a number of entries equal to
+    the number of unique values for the sensitive feature, multiplied by
+    the number of unique values of the `Y` array, multiplied by two (for
+    the Lagrange multipliers for positive and negative constraints).
+
+    With these definitions, the :meth:`signed_weights` method
+    will calculate the costs like how it is calculated in Example 4 of
```
Comment (riedgar-ms) [inline fairlearn/reductions/_moments/conditional_selection_rate.py]: Minor thing but how about "will calculate the costs similarly to Example 4 ….. but only for all `y=1` samples "

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 2 above)
Comment (romanlutz) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: There are likely more things that won't work for multi-class, but setting this up the right way to start with should prevent some future pains ;-)
Still says filter_value=0 above, though?

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 7 above)
Comment (romanlutz) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: No worries!

File: .gitignore [hunk 12]
```diff
@@ -114,3 +114,4 @@ docbuild/
 
 # generated perf test scripts
 test/perf/scripts/**
+.idea
```
Comment (romanlutz) [inline .gitignore]: nit: can you make this a new section with a comment on what .idea files are? I presume they're related to your IDE in some way?

Comment (hannanabdul55) [reply .gitignore]: Yes, it is related to the IDE I use. should I remove it altogether and just keep it in my local instead?

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 11 above)
Comment (hannanabdul55) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: Yes, I have written that in line number 186 below. I can rephrase it to this though!

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 2 above)
Comment (hannanabdul55) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: Yes, this is the value that has to be filtered. Oh sorry, I think I misinterpreted the earlier comments then, Should I rename this parameter to `good_value`? I can do that!

Comment (riedgar-ms) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: If we're going to call this true_positive_parity or similar, then `positive_value` might be better (also sounds less like a value judgement)

File: .gitignore (hunk 12 above)
Comment (riedgar-ms) [reply .gitignore]: I imagine we will acquire these over time anyway. I think a comment line above to note that this is to ignore files created by IDEA would be fine. We already have things for VS and VS Code, after all.

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 8 above)
Comment (hannanabdul55) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: Done 👍

File: .gitignore (hunk 12 above)
Comment (hannanabdul55) [reply .gitignore]: Done 👍

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 2 above)
Comment (hannanabdul55) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: Yeah, we could call this `positive_value`. I thought `filter_value` is generic and could be used for positive as well as, possibly, negative values too. Anyways, I can change it to `positive_value`. Let me know if I should.

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 5 above)
Comment (MiroDudik) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: Actually, let's rename `EqualOpportunity` -> `TruePositiveRateDifference` to match the [metric API](https://github.com/fairlearn/fairlearn-proposals/blob/master/api/METRICS.md). The same needs to propagate to the above and to the doc string below. There you can mention that this criterion is also referred to as "equal opportunity".

File: fairlearn/reductions/_moments/conditional_selection_rate.py [hunk 13]
```diff
@@ -33,13 +33,21 @@ def default_objective(self):
         """Return the default objective for moments of this kind."""
         return ErrorRate()
 
-    def load_data(self, X, y, event=None, **kwargs):
+    def load_data(self, X, y, event=None, filter_value=None, **kwargs):
         """Load the specified data into this object.
```
Comment (MiroDudik) [inline fairlearn/reductions/_moments/conditional_selection_rate.py]: I think that the following proposal will be better:
* let's get rid of the `filter_value` field
* if `event` contains `numpy.nan` those values should be excluded from conditioning

This will be consistent with the `pandas.DataFrame.groupby` and may also lead to a more concise code; e.g.:
```
>>> import pandas as pd
>>> import numpy as np
>>> df = pd.DataFrame({"event": ['a', 'a', np.nan, 'b'], "val": [4, 3, 2, 1]})
>>> df.groupby("event").size()
event
a    2
b    1
dtype: int64
```

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 2 above)
Comment (MiroDudik) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: See my comment above. Simply enter `numpy.nan` into `event` where y=0 and keep `_LABEL + "=" + str(y)` where y=1.

File: notebooks/Exp Gradient for DP(adultl).ipynb [hunk 14]
```diff
@@ -0,0 +1,192 @@
+{
```
Comment (MiroDudik) [inline notebooks/Exp Gradient for DP(adultl).ipynb]: Make this a separate PR. The notebook should have more than just the code--it needs to provide the motivation, e.g., what does it try to accomplish? And it needs to explain the steps in it more carefully.

File: notebooks/Exp Gradient for DP(law school).ipynb [hunk 15]
```diff
@@ -0,0 +1,249 @@
+{
```
Comment (MiroDudik) [inline notebooks/Exp Gradient for DP(law school).ipynb]: See the comment for the other notebook.

File: notebooks/Exp Gradient for DP.ipynb [hunk 16]
```diff
@@ -0,0 +1,248 @@
+{
```
Comment (MiroDudik) [inline notebooks/Exp Gradient for DP.ipynb]: see the comment for the other notebook.

File: fairlearn/reductions/_moments/conditional_selection_rate.py (hunk 13 above)
Comment (romanlutz) [reply fairlearn/reductions/_moments/conditional_selection_rate.py]: So where do you set it to np.nan?

**Comment by romanlutz** [comment]
Closing since #343 replaces this.
